compiler/rustc_mir/src/transform/instcombine.rs

```diff
@@ -87,6 +93,49 @@ impl OptimizationFinder<'b, 'tcx> {
         OptimizationFinder { body, tcx, optimizations: OptimizationList::default() }
     }
 
+    fn find_deref_of_address(&mut self, rvalue: &Rvalue<'tcx>, location: Location) -> Option<()> {
+        // look for the sequence
+        // _2 = &_1;
+        // ...
+        // _5 = (*_2)
+        //
+        // which we can replace the last statement with `_5 = _1` to avoid the load of _2
+        if let Rvalue::Use(op) = rvalue {
+            let local_being_derefed = match op.place()?.as_ref() {
+                PlaceRef { local, projection: [ProjectionElem::Deref] } => Some(local),
+                _ => None,
+            }?;
+
+            let stmt_index = location.statement_index;
+            // look behind for statement that assigns the local from a address of operator
+            let statements_to_look_in =
+                self.body.basic_blocks()[location.block].statements[0..stmt_index].iter().rev();
```

Comment:
This has `O(n²)` performance

Reply:
True. I originally had it at a lookback of 3, which would be better. It's not guaranteed to optimize all cases. I'll experiment with which heuristic gives the best tradeoff